src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/DictionarySlim.cs

```diff
@@ -0,0 +1,481 @@
+
+            return entries;
+        }
+
+        public KeyEnumerable Keys => new KeyEnumerable(this);
+
+        public ValueEnumerable Values => new ValueEnumerable(this);
+
+        public void CopyTo(KeyValuePair<TKey, TValue>[] array, int index)
+        {
+            Entry[] entries = _entries;
+            int i = 0;
+            int count = _count;
+            while (count > 0)
+            {
+                Entry entry = entries[i];
+                if (entry.next > -2) // part of free list?
+                {
+                    count--;
+                    array[index++] = new KeyValuePair<TKey, TValue>(
+                        entry.key,
+                        entry.value);
+                }
+                i++;
+            }
+        }
+
+        public Enumerator GetEnumerator() => new Enumerator(this); // avoid boxing
+        IEnumerator<KeyValuePair<TKey, TValue>> IEnumerable<KeyValuePair<TKey, TValue>>.GetEnumerator() => new Enumerator(this);
+        IEnumerator IEnumerable.GetEnumerator() => new Enumerator(this);
+
+        public struct Enumerator : IEnumerator<KeyValuePair<TKey, TValue>>
+        {
+            private readonly DictionarySlim<TKey, TValue> _dictionary;
+            private int _index, _count;
+            private KeyValuePair<TKey, TValue> _current;
+
+            internal Enumerator(DictionarySlim<TKey, TValue> dictionary)
+            {
+                _dictionary = dictionary;
+                _index = 0;
+                _count = _dictionary._count;
+                _current = default;
+            }
+
+            public bool MoveNext()
+            {
+                if (_count == 0)
+                {
+                    _current = default;
+                    return false;
+                }
+
+                _count--;
+
+                while (_dictionary._entries[_index].next < -1)
+                    _index++;
+
+                _current = new KeyValuePair<TKey, TValue>(
+                    _dictionary._entries[_index].key,
+                    _dictionary._entries[_index++].value);
+                return true;
+            }
+
+            public KeyValuePair<TKey, TValue> Current => _current;
+
+            object IEnumerator.Current => _current;
+
+            void IEnumerator.Reset()
+            {
+                _index = 0;
+                _count = _dictionary._count;
+            }
+
+            public void Dispose() { }
+        }
+
+        public struct KeyEnumerable : ICollection<TKey>, IReadOnlyCollection<TKey>
+        {
+            private readonly DictionarySlim<TKey, TValue> _dictionary;
+
+            internal KeyEnumerable(DictionarySlim<TKey, TValue> dictionary)
+            {
+                _dictionary = dictionary;
+            }
+
+            public int Count => _dictionary._count;
+
+            public bool IsReadOnly => true;
+
+            void ICollection<TKey>.Add(TKey item) => throw new NotSupportedException(Strings.ReadOnly_Modification);
```

Comment:
Replace `Strings.ReadOnly_Modification` with `Strings.NotSupported_KeyCollectionSet`? Same applies to other modification methods.

Reply:
I removed that string as it did not seem to add much over "The collection is read-only."

Reply:
Okay, I was using that string in the OrderedDictionary proposal as that's the string Dictionary uses. I can switch it to using this string.